Add JS validation for MCQ/MSQ for number of options in instructorFeedbackEditPage #2765

MrFool opened the pull request:
Fixes #2765

MrFool commented:
i'll update when it's green, no need to set to review yet

LowWeiLin commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -805,17 +805,27 @@ function addMcqOption(questionNumber) {
+	var numberOfOptions = $thisRow.parent().children('div').length - 1;
+
+	if (numberOfOptions <= 2) {
+	  alert('Too little choices for Multiple-choice (single answer) question. Minimum number of options is: 2.');
```
Using alerts is probably a bad idea.
When the user tries to remove an option and there are <= 2 options, he is probably trying to clear the text and write something else. Maybe do that instead?

MrFool commented:
Sounds good, didn't think of it that way, Let me fix that along with the tests in awhile.

In that case, no need to tell the user why the action is not allowed - now that it is no longer possible to have < 2 options, there is no longer a way to see the backend generated error message.

MrFool commented:
Removed alert, test cases green, with some edits to rubric test cases because it tried to delete past 2 rows to 1 row at some points.

Still thinking about how to show the "can't be done" message if we clear the text box after clicking and there is less than 2 rows, thoughts?

MrFool commented:
By the way, 

![screen shot 2015-04-01 at 3 51 58 pm](https://cloud.githubusercontent.com/assets/6291947/6936644/118230fa-d887-11e4-8a7e-1aeb332d9051.png)

![screen shot 2015-04-01 at 3 53 38 pm](https://cloud.githubusercontent.com/assets/6291947/6936682/631b2f5c-d887-11e4-9620-11f90621c4bc.png)


MCQ using MSQ HTML, intended or? If it's not intended, should open new issue?

I fixed the JS to work because one of them was being attached the wrong deletion event handler, but now it's confusing because their HTML has been intertwined before I started on the issue.

LowWeiLin commented:
Isn't that correct? the top one is Multiple-choice (multiple answers) which is MSQ, while the bottom one is Multiple choice (single answer) which is MCQ.

LowWeiLin commented:
In that case, no need to tell the user why the action is not allowed - now that it is no longer possible to have < 2 options, there is no longer a way to see the backend generated error message.
>> Backend messages can still be seen if the user meddles with the javascript. Also, it might be better to only clear the fields if there is only one option left, as users might want to re-order the options.

Still thinking about how to show the "can't be done" message if we clear the text box after clicking and there is less than 2 rows, thoughts?
>> Probably no need to show any message. It should be enough that something happens(text gets cleared)

MrFool commented:
Right, I messed up MCQ and MSQ's naming.

I got one more question though - if a user can't have less than 2 options anymore, how would we get to 'only one option'?

LowWeiLin commented:
I'm trying to say that the minimum number of options on the UI should be 1 (but back-end still checks for min of 2). The main point of this issue is to prevent 0 options on the UI, which looks confusing.

MrFool commented:
I see I see, understood, how about for rubric? When I forked it, it says check for 2 and 2, is it still 2 and 2, or 1 by 1 or?

MrFool commented:
by "it", I mean the comments that were marked:

//TODO: Check for 2

etc.

LowWeiLin commented:
Same as mcq, msq, there should be at least one cell(1 row, one col)

MrFool commented:
Done, now checks for <= 1 for MSQ MCQ and Rubric

JS added html for MSQ options also now correctly calls removeMsqOption rather than removeMcqOption previously

MrFool commented:
Just ran full test suite, dev green

LowWeiLin commented:
Nice. Could you also add/modify the UI tests to check for this new behaviour?

MrFool commented:
Tests added

MrFool commented:
Ok so for a quick explanation of the changes -

1. removeMcqOption / removeMsqOption - variables were declared without var keyword, added the var keyword.

2. removeMcqOption / removeMsqOption - was using == instead of === which is not recommended in JavaScript due to type coercion, fixed to use ===.

3. removeMcqOption / removeMsqOption - added checks for whether one last option is left and if so, they now clear the text field instead of removing the option entirely which leads to an MCQ / MSQ question not having an options

4. JavaScript generated MSQ HTML was incorrectly calling removeMcqOption, now fixed to call the correct removeMsqOption.

5. Made the same changes of 1-3 above for removing rubric row and rubric column.

6. Added tests for checking that the above functionalities added are not violated and passing in FeedbackRubricQuestionUiTest.java and InstructorFeedbackEditPageUiTest.java.

7. In InstructorFeedbackEditPage.java - 
  clickRemoveRubricRowLinkAndConfirm had a WebElement named addRubricRowLink -

  fixed to be named removeRubricRowLink to avoid confusion

  and

  clickRemoveRubricColLinkAndConfirm had a WebElement named addRubricColLink -

  fixed to be named removeRubricColLink to avoid confusion

LowWeiLin commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java:
```diff
@@ -329,7 +329,22 @@ private void testInputValidationForMcqQuestion() {
         
+        assertEquals(true, feedbackEditPage.isElementPresent("mcqOptionRow-0--1"));
+        feedbackEditPage.clickRemoveMcqOptionLink(0, -1);
+        assertEquals(true, feedbackEditPage.isElementPresent("mcqOptionRow-0--1"));
```
Also check that the value is empty?

LowWeiLin replied:
For the other question types too.

LowWeiLin commented on src/test/java/teammates/test/cases/ui/browsertests/FeedbackRubricQuestionUiTest.java:
```diff
@@ -304,6 +305,33 @@ private void testDeleteRubricQuestionAction() {
     }
     
+    private void testInputJsValidationForRubricQuestion() {
+        ______TS("JS validation test");
```
Add comments to make it easier to understand what is being done in the test, and what is tested.

LowWeiLin commented:
Thanks for the summary! Added some comments on the testing, after that it should be fine.

MrFool commented:
Added comments to the testing, unable to test for strings unless some overhaul is made - imo should be a new issue:

1. For the MCQs the value of the input box does not update when we type things into the field - remains as empty string even after selenium adds "choice 1" to it.

2. For Rubric, the textarea's inner HTML remains as the default until we type in new things and we click "submit"

LowWeiLin commented:
Why not able to test the strings? Have you tried creating a function similar to getMaxNumScaleBox in InstructorFeedbackEditPage.java? Seems weird that selenium does not support this.
If it can't be done, place a TODO comment there instead.

MrFool commented:
Yupp, I tried to target the webElement and do a .getText(), so the textarea's text was still the default despite it being empty on the frontend side - but after I manually click save question, it was indeed the new text that was typed in from the UI. It's not selenium's problem, it's got to do with the fact that the textarea's innerHTML does not update when the user changes the value inside the textarea.

LowWeiLin commented:
Try .getAttribute("innerText") or .getAttributes("textContent") instead of .getText()? From a quick search, some people seem to use that as an alternative. Otherwise, its fine to just add a TODO comment.

MrFool commented:
Yupp, just tested those as well, they also get the same value as getText() and since the inner text does not change when text is changed by the users, it returns the original value of

"This student participates well in online discussions."

E.g.

![screen shot 2015-04-05 at 1 06 17 am](https://cloud.githubusercontent.com/assets/6291947/6993700/02cce798-db30-11e4-93bf-0f04af9c12cf.png)
![screen shot 2015-04-05 at 1 06 01 am](https://cloud.githubusercontent.com/assets/6291947/6993699/02cc0440-db30-11e4-8ddd-a8498389b41c.png)


and for MCQ/MSQ, the original inner text / html is empty, it remains empty.

One way we can do the checks would be to bind the text inside the box with the HTML, but that requires more JavaScript than we have at the moment. We could look into that by opening another issue? But I'd think that is a major change.

MrFool commented:
I can open an issue and take a deeper look into it as to why the values / inner text is not being updated when the user changes the values in the text box / text area

LowWeiLin commented:
Okay sure, go ahead and create the issue.

LowWeiLin commented:
Just realized that Distribute points (among recipients) question type also needs this. Will create a separate issue for it.

MrFool commented:
Hmm ok 3262 shouldn't be too different from this, I'll see if I can tackle that, but 3263 might be harder considering that binding the text input to the HTML might be a lot more of an undertaking
